Improve docs (#681)
festim-dev/FESTIM: docs/source/userguide/export_post_processing.rst

===============
Post-processing
===============

Exports are added to the simulation object as a list of :class:`festim.Export` objects:

.. code-block:: python

    import festim as F

    my_model = F.Simulation()

    my_model.exports = [..., ...]

-------------------
Exporting solutions
-------------------

^^^^^^^^^^^
XDMF export
^^^^^^^^^^^

The most straightforward way to export solutions (concentrations, temperature) with FESTIM is to use the :class:`festim.XDMFExport` class.

> This class leverages the ``XDMFFile`` class of ``fenics`` and allows to export solutions in the XDMF format.

Reviewer: ```suggestion
This class leverages the ``XDMFFile`` class of ``fenics`` and allows solutions to be exported in the XDMF format.
```

.. code-block:: python

    import festim as F

    my_model = F.Simulation()
    my_model.mesh = F.MeshFromVertices([0, 1, 2, 3])
    my_model.materials = F.Material(id=1, D_0=1, E_D=0)
    my_model.boundary_conditions = [
        F.DirichletBC(surfaces=[1, 2], value=0, field="solute"),
    ]
    my_model.sources = [F.Source(value=1, volume=1, field="solute")]

    my_model.T = F.Temperature(500)

    my_model.exports = [
        F.XDMFExport(
            "solute",  # the field we want to export
            label="mobile",  # how the field will be labelled in the XDMF file
            filename="./mobile_conc.xdmf",
            checkpoint=False,  # needed in 1D
        )
    ]

    my_model.settings = F.Settings(
        absolute_tolerance=1e-10, relative_tolerance=1e-10, transient=False
    )

    my_model.initialise()
    my_model.run()

Running this should produce a file called ``mobile_conc.xdmf`` in the current directory.
The file can then be opened in Paraview or any other software that can read XDMF files.

For transient simulations, by default, :class:`festim.XDMFExport` will export the solution at each timestep.

> It is however possible to change this behaviour in order to limit the number of times the file is written to.

Reviewer: ```suggestion
It is possible to change this behaviour to limit the number of times the file is written to.
```

> By setting the ``mode`` attribute to ``10`` for example, the solution will be exported every 10 timesteps.

Reviewer: ```suggestion
By setting the ``mode`` attribute to ``10``, for example, the solution will be exported every 10 timesteps.
```

Setting it to ``last`` will export the solution only at the last timestep.

.. code-block:: python

    my_model.exports = [
        F.XDMFExport(
            "solute",
            label="mobile",
            filename="./mobile_conc.xdmf",
            checkpoint=False,
            mode=10,
        )
    ]

> In order for the XDMF file to be readable by FESTIM, the ``checkpoint`` attribute must be set to ``True``.

Reviewer: Do you mean paraveiw here?

```suggestion
The ``checkpoint`` attribute must be set to ``True`` for the XDMF file to be readable by Paraveiw.
```

Author: No I mean FESTIM

^^^^^^^^^^^^^^^
TXT export (1D)
^^^^^^^^^^^^^^^

> The ``TXTExport`` class allows to export solutions in a simple text format.

Reviewer: ```suggestion
The ``TXTExport`` class allows solutions to be exported in a simple text format.
```

It works in 1D only. For multi-dimensional problems, use the :class:`festim.XDMFExport` class instead.

.. code-block:: python

    import festim as F

    my_export = F.TXTExport(field="solute", filename="./mobile_conc.txt")

Adding this export to the simulation object will produce a file called ``mobile_conc.txt`` in the current directory.

> This file will contain the solution of the ``solute`` field at degree of freedom of the mesh and at each timestep.

Reviewer: not sure about this one

```suggestion
This file will contain the solution of the ``solute`` field at the degree of freedom of the mesh and at each timestep.
```

To only export at specific times in the simulation, use the ``times`` argument:

.. code-block:: python

    my_export = F.TXTExport(
        field="solute", filename="./mobile_conc.txt", times=[0, 1, 2, 3]
    )

^^^^^^^^^^^
Point value
^^^^^^^^^^^

If information about the solution at a specific point is needed, the :class:`festim.PointValue` class can be used.

> It is implemented as a derived quantity, see :ref:`Derived quantities` for more information. Here are a few examples:

Reviewer: ```suggestion
It is implemented as a derived quantity. See :ref:`Derived quantities` for more information. Here are a few examples:
```

.. code-block:: python

    import festim as F

    my_export = F.PointValue(field="solute", x=[0.5, 0.5, 0.5])
    my_export = F.PointValue(field="solute", x=(0.5, 0.5, 0.5))
    my_export = F.PointValue(field="solute", x=[0.5, 0.5])
    my_export = F.PointValue(field="solute", x=[0.5])
    my_export = F.PointValue(field="solute", x=0.5)

------------------
Derived quantities
------------------

In addition to exporting the actual solutions, it is possible to export derived quantities.
For instance, you may want to compute the flux of mobile particles at a given boundary.

> First you want to create a :class:`festim.DerivedQuantities` object. This will encompass all the derived quantities you want to compute.

Reviewer: ```suggestion
First, you want to create a :class:`festim.DerivedQuantities` object. This will encompass all the derived quantities you want to compute.
```

> Then you can add the derived quantities you want to compute to this object.

Reviewer: ```suggestion
Then, you can add the derived quantities you want to compute to this object.
```
